eth/protocols/eth/handlers.go

```diff
@@ -299,11 +306,23 @@ func handleBlockBodies66(backend Backend, msg Decoder, peer *Peer) error {
 	if err := msg.Decode(res); err != nil {
 		return fmt.Errorf("%w: message %v: %v", errDecode, msg, err)
 	}
+	metadata := func() interface{} {
+		var (
+			txsHashes   = make([]common.Hash, len(res.BlockBodiesPacket))
+			uncleHashes = make([]common.Hash, len(res.BlockBodiesPacket))
+		)
+		hasher := trie.NewStackTrie(nil)
+		for i, body := range res.BlockBodiesPacket {
+			txsHashes[i] = types.DeriveSha(types.Transactions(body.Transactions), hasher)
```

Comment:
if we were really really _really_ certain that the rlp parser is very strict, then we could deriveSha based off of the already rlp-encoded data inside the `res`

Reply:
The RLP encoded packet has txs and uncles bundled together, derivation needs them separately.

Reply:
I didn't mean that  you could take the data dump as a whole, but you can iterate the rlp list and dump out the rlp blobs. 
```
[
 [ ... transactions .. ]
[ ... uncles...]
]
```
But I think it may be a bit too optimistic to do that

Reply:
Hmm, parsing the received data might be a good thing to avoid strange attacks with some non-checked junk ending up in the database. That said, it might be interesting to keep the RLP and forward it to the consumer to avoid a re-serialization when writing the ancients.

Reply:
Oh. So we get RLP, decode to object form. In preprocessor, we encode to RLP again and then throw away the object form. Use the rlp:ed data to derive sha and push to db. Yes.